L'objectif de ce notebook est de procéder à une jointure entre les données du fichier "Updated_Combined_Data_Audience.csv" qui regroupe les programmes ayant fait les meilleures audiences par année avec un dataset de TMDB qui apporte davantage d'informations pour chaque film.


Vincent Lucas, 01/2025

In [ ]:
pip install Wikipedia-API

  Preparing metadata (setup.py) ... done
  Created wheel for Wikipedia-API: filename=Wikipedia_API-0.7.3-py3-none-any.whl size=14325 sha256=e9fb0b3855dd5dab6d3cc7457391113f7aae70815518d44e1a757160f9cf7322
  Stored in directory: /root/.cache/pip/wheels/39/c8/cd/ba9a385ad2f6e457226fe2a7dceaeb0c17a19346d2e78ff4be
Successfully built Wikipedia-API


In [ ]:
import pandas as pd
import os
import wikipediaapi

In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [ ]:
chemin_donnees = "/content/drive/MyDrive/PIP2025/Donnees"
path1 = os.path.join(chemin_donnees, "Meilleures audiences", "Updated_Combined_Data_Audience.csv")
path2 = os.path.join(chemin_donnees, "Données externes", "TMDb_Dataset.csv")
df = pd.read_csv(path1)
df2 = pd.read_csv(path2)
df

chaine                      nom film    date de diffusion  2023  \
0     France 2   ANTOINETTE DANS LES CEVENNES  2023-05-21 00:00:00  2023   
1     France 2                    BOITE NOIRE  2023-11-19 00:00:00  2023   
2     France 2                      LES VETOS  2023-02-19 00:00:00  2023   
3     France 2                     LA DARONNE  2023-03-19 00:00:00  2023   
4     France 2   LES AVENTURES DE RABBI JACOB  2023-10-22 00:00:00  2023   
...         ...                           ...                  ...   ...   
2206         M6               Corrina Corrina  2013-01-11 00:00:00  1999   
2207         M6               Turner et Hooch  2013-02-28 00:00:00  1999   
2208         M6                    La Mutante  2013-03-29 00:00:00  1999   
2209         M6              Opération dragon  2013-02-15 00:00:00  1999   
2210         M6      Le Boulanger de Valorgue  2013-01-21 00:00:00  1999   

     nationalité2  rang de diffusion téléspectateurs (millions)  \
0          FRANCE                  1                    5758382   
1          FRANCE                  1                    5523346   
2          FRANCE                  1                    4657500   
3          FRANCE                  1                    4197500   
4          FRANCE                 20                    4054371   
...           ...                ...                        ...   
2206   Etats-Unis                  1                       12,1   
2207   Etats-Unis                  3                       11,6   
2208   Etats-Unis                  1                        9,8   
2209   Etats-Unis                  5                        9,8   
2210       France                  6                        9,1   

     part d'audience (%)  
0                   27,5  
1                   26,6  
2                     22  
3                     20  
4                   17,9  
...                  ...  
2206                27,3  
2207                25,4  
2208                22,3  
2209                20,9  
2210                22,6  

[2211 rows x 8 columns]

In [ ]:
df2[df2['Title'] == 'Taxi']

TMDb_id  IMDB_id Title                                 Categories  \
1107     2330   152930  Taxi                       ['Action', 'Comedy']   
3073    11045   316732  Taxi  ['Action', 'Comedy', 'Thriller', 'Crime']   
7672   320006  4359416  Taxi         ['Comedy', 'Drama', 'Documentary']   

            Date  Country  Runtime  \
1107  1998-04-08       FR       86   
3073  2004-10-06       US       97   
7672  2015-04-15  Inconnu       82   

                                                   Cast  \
1107  [{'adult': False, 'gender': 2, 'id': 20666, 'k...   
3073  [{'adult': False, 'gender': 1, 'id': 15758, 'k...   
7672  [{'adult': False, 'gender': 2, 'id': 120229, '...   

                                                   Crew  First_Man_Age  ...  \
1107  [{'adult': False, 'gender': 2, 'id': 59, 'know...           37.0  ...   
3073  [{'adult': False, 'gender': 2, 'id': 11423, 'k...            NaN  ...   
7672  [{'adult': False, 'gender': 2, 'id': 120229, '...           55.0  ...   

         Revenue  Popularity   Note  Nb_notes       Collection  \
1107  44500000.0      37.415  6.627      2007  Taxi Collection   
3073  71255003.0      34.670  5.556       841              NaN   
7672         NaN       6.734  7.000       324              NaN   

                                       Films_Collection  Original  \
1107  [{'adult': False, 'backdrop_path': '/u9xtjq0oU...  Original   
3073                                                NaN  Original   
7672                                                NaN  Original   

                                               Keywords  \
1107  [{'id': 444, 'name': 'taxi'}, {'id': 916, 'nam...   
3073  [{'id': 1749, 'name': 'taxi driver'}, {'id': 2...   
7672  [{'id': 444, 'name': 'taxi'}, {'id': 1749, 'na...   

                                                 Poster Bechdel  
1107  https://image.tmdb.org/t/p/w185/egBxj3ude84the...     NaN  
3073  https://image.tmdb.org/t/p/w185/5EJ3XEuiXCoqqY...     3.0  
7672  https://image.tmdb.org/t/p/w185/pZf8wgEZOpTKte...     3.0  

[3 rows x 22 columns]

In [ ]:
df_head = df.head(50)

In [ ]:
found, not_found, existe_pas = 0, 0, 0
wiki_fr = wikipediaapi.Wikipedia(
    language='fr',
    user_agent='MonScript/1.0 (https://votre-site.com/contact)'
)

for index, row in df.iterrows():
  print(f'\r{index}/{len(df)}\t\t',end='')
  titre_fr = row['nom film']
  page_fr = wiki_fr.page(titre_fr)

  if page_fr.exists():
      if 'en' in page_fr.langlinks:
          page_en = page_fr.langlinks['en']
          titre_en = page_en.title
          found += 1
      else:
          not_found += 1
  else:
      existe_pas += 1

print(found, not_found, existe_pas)

2210/2211		557 92 1562


Stats pour cette méthode:
Trouvés en français: 557
Trouvés en anglais: 92
Pas trouvés (page fr): 1562

In [ ]:
import requests
import wikipediaapi

# Fonction pour effectuer une recherche sur Wikipédia
def search_wikipedia(query, lang="fr"):
    URL = f"https://{lang}.wikipedia.org/w/api.php"
    params = {
        "action": "opensearch",
        "search": query,
        "limit": 1,         # On prend le 1er résultat
        "namespace": 0,
        "format": "json"
    }
    headers = {
        "User-Agent": "MonScript/1.0 (https://votre-site.com/contact)"
    }
    response = requests.get(URL, params=params, headers=headers)
    data = response.json()
    if len(data) >= 2 and data[1]:
        return data[1][0]  # Le premier titre trouvé
    return None

# Initialisation de l'instance Wikipedia pour le français
wiki_fr = wikipediaapi.Wikipedia(
    language='fr',
    user_agent='MonScript/1.0 (https://votre-site.com/contact)'
)

# Supposons que df est votre DataFrame et que la colonne contenant le nom du film est "nom film"
found = 0
not_found = 0

for index, row in df.iterrows():
    titre_fr = row['nom film']

    # Construire la requête en simulant la saisie manuelle
    # Vous pouvez essayer différentes concaténations, par exemple "NomDuFilm film" ou "NomDuFilm (film)"
    requete = f"{titre_fr} film"
    titre_recherche = search_wikipedia(requete, lang="fr")

    if titre_recherche:
        page_fr = wiki_fr.page(titre_recherche)
        if page_fr.exists():
            # Traitement si la page a été trouvée
            found += 1
            # Optionnel : enregistrer le titre trouvé dans le DataFrame
            df.at[index, 'Titre_wiki'] = page_fr.title
            print(f"Film trouvé : {page_fr.title}")
        else:
            not_found += 1
            print(f"La page pour '{titre_fr}' (recherche: '{titre_recherche}') n'existe pas.")
    else:
        not_found += 1
        print(f"Aucun résultat trouvé pour la recherche: '{requete}'.")

print("Nombre de pages trouvées :", found)
print("Nombre de pages non trouvées :", not_found)


Aucun résultat trouvé pour la recherche: 'ANTOINETTE DANS LES CEVENNES film'.
Film trouvé : Boîte noire (film)
Aucun résultat trouvé pour la recherche: 'LES VETOS film'.
Aucun résultat trouvé pour la recherche: 'LA DARONNE film'.
Aucun résultat trouvé pour la recherche: 'LES AVENTURES DE RABBI JACOB film'.
Aucun résultat trouvé pour la recherche: 'MOURIR PEUT ATTENDRE film'.
Aucun résultat trouvé pour la recherche: 'BAC NORD film'.
Aucun résultat trouvé pour la recherche: 'LA FOLIE DES GRANDEURS film'.
Aucun résultat trouvé pour la recherche: 'A COUTEAUX TIRES film'.
Aucun résultat trouvé pour la recherche: 'JACK REACHER : NEVER GO BACK film'.
Aucun résultat trouvé pour la recherche: 'LA FILLE AU BRACELET film'.
Aucun résultat trouvé pour la recherche: 'QUI M'AIME ME SUIVE film'.
Film trouvé : L'Enquête corse (film)
Aucun résultat trouvé pour la recherche: 'MES HEROS film'.
Aucun résultat trouvé pour la recherche: 'LA FRENCH film'.
Film trouvé : La Vache (film, 2016)
Film trouvé : La S

Nombre de pages trouvées : 907
Nombre de pages non trouvées : 1304

In [ ]:
import requests
import wikipediaapi

# Fonction pour effectuer une recherche sur Wikipédia
def search_wikipedia(query, lang="fr"):
    URL = f"https://{lang}.wikipedia.org/w/api.php"
    params = {
        "action": "opensearch",
        "search": query,
        "limit": 1,         # On prend le 1er résultat
        "namespace": 0,
        "format": "json"
    }
    headers = {
        "User-Agent": "MonScript/1.0 (https://votre-site.com/contact)"
    }
    response = requests.get(URL, params=params, headers=headers)
    data = response.json()
    if len(data) >= 2 and data[1]:
        return data[1][0]  # Le premier titre trouvé
    return None

# Initialisation de l'instance Wikipedia pour le français
wiki_fr = wikipediaapi.Wikipedia(
    language='fr',
    user_agent='MonScript/1.0 (https://votre-site.com/contact)'
)


found = 0
not_found = 0

for index, row in df.iterrows():
    print(f'\r{index}/{len(df)}\t\t',end='')
    titre_fr = row['nom film']


    requete = f"{titre_fr}"
    titre_recherche = search_wikipedia(requete, lang="fr")

    if titre_recherche:
        page_fr = wiki_fr.page(titre_recherche)
        if page_fr.exists():
            # Traitement si la page a été trouvée
            found += 1
            # Optionnel : enregistrer le titre trouvé dans le DataFrame
            df.at[index, 'Titre_wiki'] = page_fr.title
        else:
            not_found += 1
    else:
        not_found += 1

print("Nombre de pages trouvées :", found)
print("Nombre de pages non trouvées :", not_found)


2210/2211		Nombre de pages trouvées : 2117
Nombre de pages non trouvées : 94


Nombre de pages trouvées : 2117
Nombre de pages non trouvées : 94

Stratégie:
On parcourt tous les films du df des meilleures audiences
Pour chaque film:
- Si on le retrouve dans les données externes, on récupère cette ligne
- Sinon on le traduit en anglais et on récupère cette ligne


In [ ]:
import requests
import wikipediaapi

# Fonction pour effectuer une recherche sur Wikipédia
def search_wikipedia(query, lang="fr"):
    URL = f"https://{lang}.wikipedia.org/w/api.php"
    params = {
        "action": "opensearch",
        "search": query,
        "limit": 1,         # On prend le 1er résultat
        "namespace": 0,
        "format": "json"
    }
    headers = {
        "User-Agent": "MonScript/1.0 (https://votre-site.com/contact)"
    }
    response = requests.get(URL, params=params, headers=headers)
    data = response.json()
    if len(data) >= 2 and data[1]:
        return data[1][0]
    return None


wiki_fr = wikipediaapi.Wikipedia(
    language='fr',
    user_agent='MonScript/1.0 (https://votre-site.com/contact)'
)

df_final = pd.DataFrame(columns=df2.columns)

found = 0
not_found = 0

for index, row in df.iterrows():
    print(f'\r{index}/{len(df)}\t\t',end='')
    titre_fr = row['nom film']

    matching_rows = df2[df2['Title'] == titre_fr]
    if not matching_rows.empty:
        df_final = pd.concat([df_final, matching_rows.iloc[[0]]], ignore_index=True)
    else:
        requete = f"{titre_fr}"
        titre_recherche = search_wikipedia(requete, lang="fr")

        if titre_recherche:
            page_fr = wiki_fr.page(titre_recherche)
            if page_fr.exists():
              if 'en' in page_fr.langlinks:
                  page_en = page_fr.langlinks['en']
                  titre_en = page_en.title
                  matching_rows = df2[df2['Title'] == titre_en]
                  if not matching_rows.empty:
                      df_final = pd.concat([df_final, matching_rows.iloc[[0]]], ignore_index=True)
                      found += 1
            else:
                not_found += 1
        else:
            not_found += 1

print("Nombre de pages trouvées :", found)
print("Nombre de pages non trouvées :", not_found)



0/2211		

<ipython-input-30-fb02023c96f4>:53: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  df_final = pd.concat([df_final, matching_rows.iloc[[0]]], ignore_index=True)


2210/2211		Nombre de pages trouvées : 1761
Nombre de pages non trouvées : 94


In [ ]:
df_final

TMDb_id   IMDB_id                              Title  \
0     669664  11013434            My Donkey, My Lover & I   
1        760     69747  The Mad Adventures of Rabbi Jacob   
2     370172   2382320                     No Time to Die   
3     343611   3393786        Jack Reacher: Never Go Back   
4      39386    380349                  The Corsican File   
...      ...       ...                                ...   
1047     114    100405                       Pretty Woman   
1048   37652    118976                             Didier   
1049   24260     80929                  Inspector Blunder   
1050    6951     98536                     Turner & Hooch   
1051    9461     70034                   Enter the Dragon   

                                             Categories        Date  Country  \
0                    ['Comedy', 'Adventure', 'Romance']  2020-09-16       FR   
1                                            ['Comedy']  1973-10-17  Inconnu   
2                   ['Adventure', 'Action', 'Thriller']  2021-09-29       US   
3                                ['Action', 'Thriller']  2016-10-19       US   
4                                            ['Comedy']  2004-10-06       FR   
...                                                 ...         ...      ...   
1047                              ['Romance', 'Comedy']  1990-03-23       US   
1048                                         ['Comedy']  1997-01-28       FR   
1049                                         ['Comedy']  1980-12-03       FR   
1050  ['Action', 'Comedy', 'Thriller', 'Crime', 'Fam...  1989-07-28  Inconnu   
1051                                         ['Action']  1973-08-17       US   

     Runtime                                               Cast  \
0         97  [{'adult': False, 'gender': 1, 'id': 983929, '...   
1        100  [{'adult': False, 'gender': 2, 'id': 11187, 'k...   
2        163  [{'adult': False, 'gender': 2, 'id': 8784, 'kn...   
3        118  [{'adult': False, 'gender': 2, 'id': 500, 'kno...   
4         92  [{'adult': False, 'gender': 2, 'id': 28781, 'k...   
...      ...                                                ...   
1047     119  [{'adult': False, 'gender': 2, 'id': 1205, 'kn...   
1048     104  [{'adult': False, 'gender': 2, 'id': 28281, 'k...   
1049     100  [{'adult': False, 'gender': 2, 'id': 11214, 'k...   
1050     100  [{'adult': False, 'gender': 2, 'id': 31, 'know...   
1051     102  [{'adult': False, 'gender': 2, 'id': 19429, 'k...   

                                                   Crew  First_Man_Age  ...  \
0     [{'adult': False, 'gender': 1, 'id': 7296, 'kn...            NaN  ...   
1     [{'adult': False, 'gender': 2, 'id': 1657, 'kn...           59.0  ...   
2     [{'adult': False, 'gender': 2, 'id': 947, 'kno...           53.0  ...   
3     [{'adult': False, 'gender': 2, 'id': 500, 'kno...           54.0  ...   
4     [{'adult': False, 'gender': 2, 'id': 2949, 'kn...           52.0  ...   
...                                                 ...            ...  ...   
1047  [{'adult': False, 'gender': 2, 'id': 376, 'kno...           41.0  ...   
1048  [{'adult': False, 'gender': 2, 'id': 4275, 'kn...           46.0  ...   
1049  [{'adult': False, 'gender': 2, 'id': 1657, 'kn...           36.0  ...   
1050  [{'adult': False, 'gender': 2, 'id': 598, 'kno...           33.0  ...   
1051  [{'adult': False, 'gender': 2, 'id': 9217, 'kn...           33.0  ...   

          Revenue  Popularity   Note  Nb_notes               Collection  \
0             NaN       8.793  6.478       250                      NaN   
1             NaN      10.348  7.400       796                      NaN   
2     774153007.0     161.005  7.458      4538    James Bond Collection   
3     162146076.0      80.498  5.898      4131  Jack Reacher Collection   
4             NaN       6.546  5.940       298                      NaN   
...           ...         ...    ...       ...                      ...   
1047  463406268.0      44.467  7

In [ ]:
path = os.path.join(chemin_donnees, "Lucas.csv")
df_final.to_csv(path, index=False)